### PR TITLE
[f] OUT-228 - Using Missive rules if a comment unsub

### PR DESCRIPTION
**Pull request by Dat-H-Tran**
[https://linear.app/pdw/issue/OUT-228/using-missive-rules-if-a-comment-of-exactly-unsubscribe-or-stop-case](https://linear.app/pdw/issue/OUT-228/using-missive-rules-if-a-comment-of-exactly-unsubscribe-or-stop-case)
[https://linear.app/pdw/issue/OUT-245/implement-api-for-admins-comment-resubscribe-and-unsubscribe](https://linear.app/pdw/issue/OUT-245/implement-api-for-admins-comment-resubscribe-and-unsubscribe)

### CODE REVIEW

File: backend/services/BroadcastService.ts [hunk 1]
```diff
@@ -420,6 +420,25 @@ const isBroadcastRunning = async (): Promise<boolean> => {
...
+    .set({ unsubscribed: isUnsubscribe })
+    .where(eq(authors.phoneNumber, phoneNumber));
+
+  let postMessage = ''
+  if (isUnsubscribe) {
+    postMessage = `This phone number ${phoneNumber} has now been unsubscribed by ${authorName}.`;
+  } else {
+    postMessage = `This phone number ${phoneNumber} has now been resubscribed by ${authorName}.`;
+  }
+  return await MissiveUtils.createPost(postMessage);
```
Comment (kyphan1) [inline backend/services/BroadcastService.ts:439]: It should handle the response result here.

I think it's best for MissiveUtils to log the response body for debugging later.

File: backend/services/BroadcastService.ts [hunk 2]
```diff
@@ -420,6 +420,25 @@ const isBroadcastRunning = async (): Promise<boolean> => {
...
+    .update(authors)
+    .set({ unsubscribed: isUnsubscribe })
+    .where(eq(authors.phoneNumber, phoneNumber));
+
+  let postMessage = ''
+  if (isUnsubscribe) {
+    postMessage = `This phone number ${phoneNumber} has now been unsubscribed by ${authorName}.`;
+  } else {
+    postMessage = `This phone number ${phoneNumber} has now been resubscribed by ${authorName}.`;
+  }
```
Comment (kyphan1) [inline backend/services/BroadcastService.ts:438]: Why do we need them? The response message should be handled by the Controller.

File: backend/middlewares/missiveWebhookCallbackVerify.ts [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+import * as DenoSentry from 'sentry/deno'
+import { decodeHex } from 'encoding/hex.ts'
+import AppResponse from "../misc/AppResponse.ts";
+
+const missiveWebhookCallbackVerify = async (req: Request, res: Response, next: NextFunction) => {
+  const requestHeaderSig = req.headers['x-hook-signature'];
+
+  if (!requestHeaderSig || typeof requestHeaderSig !== 'string') {
+    DenoSentry.captureException('Bad Request: X-Hook-Signature header is missing or invalid');
+    return res.status(401).json({ error: 'Missing or invalid authentication header' });
```
Comment (kyphan1) [inline backend/middlewares/missiveWebhookCallbackVerify.ts:12]: Can extend `AppResponse.unauthorized(res)` to accept the message.

File: backend/middlewares/missiveWebhookCallbackVerify.ts [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
+  try {
+    const verified = await verifySignature(requestHeaderSig, req.body);
+    if (!verified) {
+      DenoSentry.captureException('Signature not match')
+      return AppResponse.unauthorized(res)
+    }
+    next();
+  } catch (error) {
+    log.error('Error during signature verification:', error)
```
Comment (kyphan1) [inline backend/middlewares/missiveWebhookCallbackVerify.ts:23]: Why not log to Sentry? 🤔

File: backend/lib/Missive.ts [hunk 5]
```diff
@@ -29,7 +31,31 @@ const sendMessage = (message: string, toPhone: string) => {
...
+
+  const response = await fetch(CREATE_POST_URL, {
+    method: 'POST',
+    headers: headers,
+    body: JSON.stringify(postData),
+  });
+
+  if (!response.ok) {
+    throw new Error(`HTTP error! status: ${response.status}`);
+  }
```
Comment (kyphan1) [inline backend/lib/Missive.ts:52]: Should print to stderr

File: backend/controllers/BroadcastController.ts [hunk 6]
```diff
@@ -76,11 +77,47 @@ async function updateTwilioStatus(req: Request, res: Response) {
...
+    isUnsubscribe = false;
+  } else {
+    return AppResponse.badRequest(res, 'Invalid request path');
+  }
+
+  try {
+    await BroadcastService.updateSubscriptionStatus(phoneNumber, isUnsubscribe, authorName);
+    return AppResponse.ok(res, {
+      message: `Author ${isUnsubscribe ? 'unsubscribed' : 'resubscribed'} successfully`
+    }, 200);
```
Comment (kyphan1) [inline backend/controllers/BroadcastController.ts:108]:
```suggestion
    return AppResponse.ok(res, {
      message: `Author ${isUnsubscribe ? 'unsubscribed' : 'resubscribed'} successfully`
    });
```

File: backend/services/BroadcastService.ts (hunk 2 above)
Comment (Dat-H-Tran) [reply backend/services/BroadcastService.ts]: it's logic, even though simple. The controller should not be handling these.

File: backend/controllers/BroadcastController.ts [hunk 7]
```diff
@@ -76,11 +77,50 @@ async function updateTwilioStatus(req: Request, res: Response) {
...
+    return AppResponse.badRequest(res, 'Invalid request path')
+  }
+
+  try {
+    await BroadcastService.updateSubscriptionStatus(phoneNumber, isUnsubscribe, authorName)
+    return AppResponse.ok(res, {
+      message: `Author ${isUnsubscribe ? 'unsubscribed' : 'resubscribed'} successfully`,
+    }, 200)
+  } catch (error) {
+    console.error('Error in commentChangeSubscription:', error)
```
Comment (kyphan1) [inline backend/controllers/BroadcastController.ts:113]:
```suggestion
    logger.error('Error in commentChangeSubscription:', error)
```
You should print the `phoneNumber, isUnsubscribe, authorName` out

File: backend/controllers/BroadcastController.ts [hunk 8]
```diff
@@ -76,11 +77,50 @@ async function updateTwilioStatus(req: Request, res: Response) {
   return AppResponse.ok(res, {}, 204)
 }
 
+async function commentChangeSubscription(req: Request, res: Response) {
```
Comment (kyphan1) [inline backend/controllers/BroadcastController.ts:80]: It should have a better name to indicate you're only targeting `Unsubscribe` or `Resubscribe`

File: backend/controllers/BroadcastController.ts [hunk 9]
```diff
@@ -76,11 +78,50 @@ async function updateTwilioStatus(req: Request, res: Response) {
...
+    return AppResponse.badRequest(res, 'Invalid request path')
+  }
+
+  try {
+    await BroadcastService.updateSubscriptionStatus(phoneNumber, isUnsubscribe, authorName)
+    return AppResponse.ok(res, {
+      message: `Author ${isUnsubscribe ? 'unsubscribed' : 'resubscribed'} successfully`,
+    }, 200)
+  } catch (error) {
+    log.error('Error in commentChangeSubscription:', error)
```
Comment (kyphan1) [inline backend/controllers/BroadcastController.ts:114]:
```suggestion
    log.error('Error in commentChangeSubscription: ${error.message}. Data: ${phoneNumber}, ${isUnsubscribe}, ${authorName}')
```

File: backend/lib/Missive.ts [hunk 10]
```diff
@@ -29,7 +33,32 @@ const sendMessage = (message: string, toPhone: string) => {
...
+
+  const response = await fetch(CREATE_POST_URL, {
+    method: 'POST',
+    headers: headers,
+    body: JSON.stringify(postData),
+  })
+
+  if (!response.ok) {
+    log.error(`HTTP error! status: ${response.status}`)
```
Comment (kyphan1) [inline backend/lib/Missive.ts:53]:
```suggestion
    log.error(`HTTP error! status: ${response.status} ${response.body}`)
```

File: backend/misc/AppResponse.ts [hunk 11]
```diff
@@ -13,6 +15,14 @@ const ok = (res: Response, body = {}, code = 200) => {
   return res.status(code).json(body)
 }
 
+const badRequest = (res: Response, errorMessage: string = BAD_REQUEST_ERR) => {
+  return res.status(400).json({ message: errorMessage })
+}
+
+const internalServerError = (res: Response, errorMessage: string = INTERNAL_SERVER_ERR) => {
+  return res.status(500).json({ message: errorMessage })
+}
```
Comment (kyphan1) [inline backend/misc/AppResponse.ts:24]: We should log to Slack here

File: backend/services/BroadcastService.ts [hunk 12]
```diff
@@ -420,6 +420,30 @@ const isBroadcastRunning = async (): Promise<boolean> => {
...
+    .update(authors)
+    .set({ unsubscribed: isUnsubscribe })
+    .where(eq(authors.phoneNumber, phoneNumber))
+
+  let postMessage = ''
+  if (isUnsubscribe) {
+    postMessage = `This phone number ${phoneNumber} has now been unsubscribed by ${authorName}.`
+  } else {
+    postMessage = `This phone number ${phoneNumber} has now been resubscribed by ${authorName}.`
+  }
```
Comment (kyphan1) [inline backend/services/BroadcastService.ts:438]:
```suggestion
  const action = isUnsubscribe ? 'unsubscribed' : 'resubscribed'
  const postMessage = `This phone number ${phoneNumber} has now been ${action} by ${authorName}.`
```

File: backend/services/BroadcastService.ts [hunk 13]
```diff
@@ -420,6 +420,30 @@ const isBroadcastRunning = async (): Promise<boolean> => {
...
+  try {
+    return await MissiveUtils.createPost(postMessage)
+  } catch (error) {
+    log.error('Failed to create post:', error)
```
Comment (kyphan1) [inline backend/services/BroadcastService.ts:442]:
```suggestion
    log.error('Failed to create post: ${error.message}')
```

**Review by kyphan1 — APPROVED** [review]

**Review by kyphan1 — CHANGES_REQUESTED** [review]
Missive key should be updated reactively

**Review by kyphan1 — COMMENTED** [review]
LGTM

**Review by kyphan1 — APPROVED** [review]

**Pull request merged**